File: Ch02/ex2_26.erl
-module(ex2_26).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: 2.26 - playing with append/cons/list").
-export([append/2, cons/2, list/2]).
-include_lib("eunit/include/eunit.hrl").

append(L1, L2)  -> L1 ++ L2.
cons(L1, L2)    -> [ L1 | L2 ].
list(L1, L2)    -> [ L1, L2 ].

%%% TESTS

append_test() ->
    ?assert([1, 2, 3, 4, 5, 6] =:= append([1, 2, 3], [4, 5, 6])).

cons_test() ->
    ?assert([[1, 2, 3], 4, 5, 6] =:= cons([1, 2, 3], [4, 5, 6])).

list_test() ->
    ?assert([[1, 2, 3], [4, 5, 6]] =:= list([1, 2, 3], [4, 5, 6])).

File: Ch02/ex2_26.exs
# SICP in Elixir: Exercise 2.26 - playing with append/cons/list

defmodule Ex2_26 do
  def append(x, y), do: x ++ y
  def cons(x, y), do: [x | y]
  def list(x, y), do: [x, y]
end

x = [1, 2, 3]
y = [4, 5, 6]

result = Ex2_26.append(x, y)
unless result == [1, 2, 3, 4, 5, 6], do: throw(result)

result = Ex2_26.cons(x, y)
unless result == [[1, 2, 3], 4, 5, 6], do: throw(result)

result = Ex2_26.list(x, y)
unless result == [[1, 2, 3], [4, 5, 6]], do: throw(result)
